src/qibo/quantum_info/metrics.py

```diff
@@ -229,60 +226,68 @@ def fidelity(state, target, check_hermitian: bool = False, backend=None):
 
     # check purity if both states are density matrices
     if len(state.shape) == 2 and len(target.shape) == 2:
-        purity_state = purity(state)
-        purity_target = purity(target)
+        purity_state = purity(state, backend=backend)
+        purity_target = purity(target, backend=backend)
 
         # if both states are mixed, default to full fidelity calculation
         if (
             abs(purity_state - 1) > PRECISION_TOL
             and abs(purity_target - 1) > PRECISION_TOL
         ):
             # using eigh since rho is supposed to be Hermitian
-            if check_hermitian is False or _check_hermitian_or_not_gpu(
-                state, backend=backend
-            ):
-                eigenvalues, eigenvectors = np.linalg.eigh(state)
-            else:
-                eigenvalues, eigenvectors = np.linalg.eig(state)
+            eigenvalues, eigenvectors = backend.calculate_eigenvectors(
+                state,
+                hermitian=(
+                    check_hermitian is False
+                    or _check_hermitian_or_not_gpu(state, backend=backend)
+                ),
```

Comment:
I think it's best to put this `bool` in a variable first just for the sake of code clarity